[JIT] memory planning base (no strategies)

makslevental opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #63873
* #63872
* #63539
* #63538
* #63537
* #63835
* __->__ #62174
* #62995

This PR adds static memory planning for traces with shape info. The three strategies implemented so far are stacked on top. They are:

* Linear scan (a heuristic based on https://www.usenix.org/legacy/events/vee05/full_papers/p132-wimmer.pdf)
* Greedy by size (a heuristic that allocates largest tensors first)
* Greedy by operator breadth (a heuristic that allocates largest tensors for operators with largest breadths first)

The latter 2 are based on https://arxiv.org/pdf/2001.03288.pdf. 

Most of the action is under `torch/csrc/jit/passes/memory_planning/`. 
Currently this only supports CPU. In principle there's nothing blocking supporting GPU (just requires adding cudaAllocator to GetAllocator) but there are lots of implicit assumptions encoded in the allocation strategies that are probably false for GPU.

Currently the tests don't do anything except exercise the code paths. As we find corner cases we will add tests to check for them.

TODO: add naive strategy to this PR.

Commit:
first sketch commit for memory planning

[ghstack-poisoned]

makslevental commented on aten/src/ATen/core/function_schema.cpp:
```diff
@@ -2,7 +2,7 @@
 namespace c10 {
 
-void FunctionSchema::dump() const {
+TORCH_API void FunctionSchema::dump() const {
```
necessary for debugging schema. 

question: why is this in it's own cpp?

makslevental commented on aten/src/ATen/core/type.cpp:
```diff
@@ -1012,8 +1012,6 @@ TensorType::TensorType(
       undefined_(undefined) {}
 
 TensorTypePtr TensorType::create(const at::Tensor& t) {
-  VaryingShape<bool> contiguity;
-  VaryingShape<size_t> stride_indices;
```
not used

makslevental commented on torch/csrc/jit/runtime/profiling_graph_executor_impl.cpp:
```diff
@@ -27,7 +27,6 @@
 #include <torch/csrc/jit/passes/peephole.h>
 #include <torch/csrc/jit/passes/remove_expands.h>
 #include <torch/csrc/jit/passes/remove_mutation.h>
-#include <torch/csrc/jit/passes/requires_grad_analysis.h>
```
not used

makslevental commented on torch/csrc/jit/runtime/profiling_record.cpp:
```diff
@@ -3,8 +3,6 @@
 #include <ATen/core/interned_strings.h>
 #include <c10/util/irange.h>
 #include <torch/csrc/jit/jit_log.h>
-#include <torch/csrc/jit/passes/clear_profiling.h>
-#include <torch/csrc/jit/passes/constant_propagation.h>
```
not used

makslevental commented on torch/csrc/jit/runtime/register_prim_ops_fulljit.cpp:
```diff
@@ -2,23 +2,16 @@
+#include <torch/csrc/jit/runtime/static/impl.h>
 
 #include <algorithm>
-#include <bitset>
-#include <cctype>
```
not used (same below)

makslevental commented on torch/csrc/jit/runtime/profiling_graph_executor_impl.h:
```diff
@@ -50,5 +50,7 @@ struct ProfilingGraphExecutorImpl : public GraphExecutorImplBase {
   c10::optional<size_t> remaining_bailout_depth_;
 };
 
+TORCH_API void runMemoryPlanningPipeline(std::shared_ptr<Graph>&, std::unique_ptr<ProfilingRecord>&);
+
```
exposed just for testing ease

makslevental commented on torch/csrc/jit/runtime/register_prim_ops_fulljit.cpp:
```diff
@@ -2,23 +2,16 @@
 
 #include <ATen/core/ivalue.h>
 #include <c10/util/irange.h>
+#include <torch/csrc/jit/runtime/static/impl.h>
```
this might create a circular include dependency?

Commit:
Update on "first sketch commit for memory planning"


this is just a sketch to see if i have all of the relevant pieces identified and the rough structure understood correctly. all brutal comments very welcome!

can test like this 

https://gist.github.com/makslevental/3cf89e4221aeb37c6f10aacfbb9ceb52

debug output looks like this 

https://gist.github.com/makslevental/70ae9c2079e8126ae81d7a6587f9a37e

[ghstack-poisoned]

makslevental commented on torch/csrc/jit/runtime/profiling_graph_executor_impl.h:
```diff
@@ -50,5 +50,7 @@ struct ProfilingGraphExecutorImpl : public GraphExecutorImplBase {
   c10::optional<size_t> remaining_bailout_depth_;
 };
 
+TORCH_API void runMemoryAllocationPipeline(std::shared_ptr<Graph>&, std::unique_ptr<ProfilingRecord>&);
+
```
just for testing

makslevental commented on torch/csrc/jit/runtime/profiling_record.cpp:
```diff
@@ -200,6 +198,49 @@ void ProfilingRecord::insertShapeProfile(Node* n, size_t offset) {
+  auto pn = createProfileNode(nullptr, {o});
+  pn->ty_(attr::profiled_type, TensorType::get());
+
+  std::function<void(Stack&)> memory_profiler = [this, o](Stack& stack) {
+    int64_t frame_id = 0;
+    pop(stack, frame_id);
+    IValue v;
+    pop(stack, v);
+    if (v.isTensor()) {
+      std::lock_guard<std::mutex> lock(this->mutex_);
+      auto& bytes_per_t = nbytes_per_frame_[frame_id];
+      auto& t = v.toTensor();
+      if (t.defined()) {
+        auto t_nbytes = t.storage().nbytes();
+        GRAPH_DEBUG(
+            "In run ",
+            frame_id,
+            " annotating %",
+            o->debugName(),
+            " computed nbytes ", t_nbytes);
+        if (bytes_per_t.count(o) == 0) {
+          bytes_per_t.insert({o, t_nbytes});
+        } else {
+          auto prev_t_bytes = bytes_per_t.at(o);
+          GRAPH_DEBUG("Existing bytes for %", o->debugName(), " ", prev_t_bytes)
+          bytes_per_t[o] = std::max(prev_t_bytes, t_nbytes);
+          GRAPH_DEBUG("Updated bytes for %", o->debugName(), " ", bytes_per_t[o])
+        }
+      } else {
+        bytes_per_t[o] = 0;
+      }
+    }
+    // passing t through
+    push(stack, v);
+  };
+
+  pn->setCallback(memory_profiler);
+  pn->insertAfter(n);
+}
+
```
this duplicates almost all of `insertShapeProfile` because `TensorType` doesn't store `nbytes`. I thought about extending `TensorType` to do that but figured that might be a big change...

Commit:
Update on "first sketch commit for memory planning"


this is just a sketch to see if i have all of the relevant pieces identified and the rough structure understood correctly. all brutal comments very welcome!

can test like this 

https://gist.github.com/makslevental/3cf89e4221aeb37c6f10aacfbb9ceb52

debug output looks like this 

https://gist.github.com/makslevental/70ae9c2079e8126ae81d7a6587f9a37e

[ghstack-poisoned]

Commit:
Update on "first sketch commit for memory planning"


this is just a sketch to see if i have all of the relevant pieces identified and the rough structure understood correctly. all brutal comments very welcome!

can test like this 

https://gist.github.com/makslevental/3cf89e4221aeb37c6f10aacfbb9ceb52

debug output looks like this 

https://gist.github.com/makslevental/70ae9c2079e8126ae81d7a6587f9a37e

[ghstack-poisoned]

Krovatkin commented on test/cpp/jit/test_memory_planning.cpp:
```diff
@@ -0,0 +1,48 @@
+  auto stack = createStack({input, hx, cx, w_ih, w_hh});
+
+  // run once to type info
+  auto pr = jit::ProfilingRecord::instrumentGraph(g);
```
are you using Profiler for prototyping or we would be also using it in the final version? I'm asking because if we will be using it in the final version we will need to figure out how we will guard if shapes/strides/device change?

makslevental replied:
I started out using Profiler for prototyping before I wrote this https://github.com/pytorch/pytorch/pull/62377. Will transition soon.

Krovatkin replied:
Yeah, this makes sense!

Krovatkin commented on torch/csrc/jit/passes/memory_planning.cpp:
```diff
@@ -0,0 +1,159 @@
+
+std::pair<size_t, std::map<Value*, c10::TensorTypePtr>>
+computeAlignedStorageSize(std::vector<Node*>& nodes) {
+  // TODO: liveness analysis
```
👍

makslevental replied:
working on it as we speak 😂

makslevental replied:
@desertfire oh great! just curious: is there an explanation somewhere on the technique?

eellison replied:
we should be extremely careful reusing code from previous projects without auditing them, bailouts/static runtime etc. the liveness linked there only does Value liveness it does not do alias analysis / handling of views etc

Krovatkin commented on torch/csrc/jit/passes/memory_planning.cpp:
```diff
@@ -0,0 +1,159 @@
+}
+
+Node* insertAllocStorageNode(std::shared_ptr<Graph>& graph, size_t total_size) {
+  auto deviceType = findDeviceType(graph);
```
would you anticipate significant changes to this design if we need to deal with more than one device type?

Krovatkin replied:
what would be the fallback if we can't allocate enough memory all at once?

makslevental replied:
>would you anticipate significant changes to this design if we need to deal with more than one device type?

no it shouldn't be too hard to initially allocate storage on `n` devices and then tell the `prim::allocateTensor` to allocate in the correct storage.

>what would be the fallback if we can't allocate enough memory all at once?

I haven't thought exactly about this but it would be the same fallback as when an op doesn't have an out variant.

Krovatkin replied:
hhm could you please elaborate? Not being able to allocate enough memory is a run-time condition, vs not having an out variant is a compile time one?

makslevental replied:
>hhm could you please elaborate? Not being able to allocate enough memory is a run-time condition, vs not having an out variant is a compile time one?

What I mean is that we should fallback to the same allocation strategy as for eager i.e. do not plan for those tensors (just as we don't plan for tensors that are produced by ops that don't have out variants).

Krovatkin commented on torch/csrc/jit/passes/memory_planning.cpp:
```diff
@@ -0,0 +1,159 @@
+  // TODO: pass device type here
+  storage->i_(attr::total_size, total_size);
+  storage->i_(attr::device, static_cast<int8_t>(deviceType));
+  storage->insertBefore(graph->nodes().front());
```
I was wondering if we should pass these as arguments to `AllocateStorage` (as constants for now) so when we will get to supporting dynamic shapes, we would just compute the sizes (instead of passing constants)?

makslevental replied:
naturally inputs vs attrs has its pros and cons: pro is what you said, con is it clutters the IR making downstream manipulation more difficult (i think?). i can the "input constant" route and compare differences.

Krovatkin replied:
I wonder if we should lean into it in some later PR. I don't know how else we would support dynamic shapes and it would be nice to flush out any potential issues as early as possible as they can affect an overall design, IMO. @desertfire  @hlu1  ?

Krovatkin commented on torch/csrc/jit/passes/memory_planning.cpp:
```diff
@@ -0,0 +1,159 @@
+  size_t total_size = storage->i(attr::total_size);
+  for (auto* node : out_nodes) {
+    auto* alloc = graph->create(prim::AllocateTensor, 1);
+    node->addInput(alloc->output());
```
looks like we are turning an op here into its out variant by adding an extra input? This is probably worth a comment or two.

makslevental replied:
this is @eellison's insight - that nodes don't own schema but actually look up schema in the registry based number and type of inputs/outputs. will add a comment for sure because it was surprising to me as well.

Krovatkin replied:
exactly, it's not super obvious :)

Krovatkin commented on torch/csrc/jit/passes/memory_planning.cpp:
```diff
@@ -0,0 +1,159 @@
+    std::tie(sizes, strides) = getSizesStrides(ttp);
+    TORCH_CHECK(offset + size <= total_size, "not enough memory");
+    alloc->i_(attr::size, size);
+    alloc->i_(attr::offset, offset);
```
do we have alignment requirements for individual tensors? 
say, the first tensor is `torch.rand(1, dtype=torch.int8)` we probably don't want the second tensor start on a byte boundary?

makslevental replied:
@ngimel I think you might know about this...

ngimel replied:
We don't have requirements other than alignment should be multiple of `sizeof(type)`, but perf will be better if alignment is better (e.g. vectorization on gpu will kick in more or less when alignment is `4*sizeof(type)`, cache lines are 128 bytes so with 128 bytes alignment you are getting better cache utilization, tensor cores can be used with, I don't remember, 256 or 512 bytes alignment?). CUDA caching allocator I think lines all new allocations to 512 bytes.

Krovatkin commented on torch/csrc/jit/passes/memory_planning.cpp:
```diff
@@ -0,0 +1,159 @@
+
+    offset += size;
+  }
+  TORCH_CHECK(offset == total_size, "exceeded total slab size");
```
this should probably be an assert

Krovatkin commented on torch/csrc/jit/passes/memory_planning.cpp:
```diff
@@ -0,0 +1,159 @@
+    alloc->i_(attr::offset, offset);
+    alloc->is_(attr::sizes, sizes);
+    alloc->is_(attr::stride, strides);
+    alloc->i_(attr::device, static_cast<int8_t>(storage->i(attr::device)));
```
again, I wonder if these should be arguments vs static attributes?

Krovatkin commented on test/cpp/jit/test_memory_planning.cpp:
```diff
@@ -0,0 +1,48 @@
+  Code cd2(pr->profiled_graph_, "");
+  InterpreterState is2{cd2};
+  stack = createStack({input, hx, cx, w_ih, w_hh});
+  is2.run(stack);
```
could you also compare the results from the profiling run with the ones from the optimized run with `allequal`?

makslevental replied:
what is `allequal`?

Krovatkin replied:
@makslevental arrgh sorry I meant allclose , https://pytorch.org/docs/stable/generated/torch.allclose.html

Krovatkin commented on torch/csrc/jit/passes/memory_planning.cpp:
```diff
@@ -0,0 +1,159 @@
+namespace jit {
+
+c10::TensorTypePtr checkIsSupported(const Value& value) {
+  TORCH_CHECK(
```
I feel like none of these conditions should trigger an error, unless you are going to catch it somewhere. memory planning is an optimization and as such should bail rather than throw

makslevental replied:
agreed. will change to something else.

Krovatkin replied:
@makslevental arrgh sorry I meant `allclose` , https://pytorch.org/docs/stable/generated/torch.allclose.html

Krovatkin commented on torch/csrc/jit/runtime/register_prim_ops_fulljit.cpp:
```diff
@@ -127,6 +130,48 @@ RegisterOperators reg(
+             push(stack, std::move(storage));
+           };
+         },
+         aliasAnalysisSpecialCase()),
```
all these will need special cases in Alias Analysis later on. Unless we are planning on doing this as the very last optimization

Krovatkin commented on torch/csrc/jit/passes/memory_planning.cpp:
```diff
@@ -0,0 +1,159 @@
+  return std::make_pair(aligned_size, managed_tensors);
+}
+
+DeviceType findDeviceType(std::shared_ptr<Graph>& graph) {
```
nitpick: TensorExprs has a similar utility function, maybe we could use either one.

makslevental replied:
is this the one? https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/tensorexpr/kernel.cpp

Krovatkin replied:
`pickDeviceType` 
https://github.com/pytorch/pytorch/blob/a9b0a921d592b328e7e80a436ef065dadda5f01b/torch/csrc/jit/tensorexpr/kernel.cpp#L160

Krovatkin reviewed (commented):
some comments

eellison reviewed (commented):
Looks great!!! I dont see anything immediately blocking landing. 

Future to-dos are:

- liveness / alias analysis / handling of views
- better memory reuse algorithm
- inplacing as an optimization
- dynamic shapes / values as arguments instead of attributes. I think the algorithmic work to do dynamic shapes memory planning etc is going to be way more than just size as attributes / inputs so I'm fine delaying that until we actually work on dynamic shapes. 
- use of strides - this isn't blocking, but we should try to see if we can get this to work without requiring striding information.  striding information is very difficult to prove statically, which makes it hard to roll out with dynamic shapes, and it's extremely difficult to reason about whether optimizations preserve striding or not. 


One other general thing we'll have to be wary of is operators that do not have out variants. When they don't have out variants, we are forking the memory allocator in two:
the one generated by our compiler, and the one used by eager backend. This could increase high-water mark of memory usage since the compiler allocator will always be allocating at maximum size. I don't think this is a hard blocker or anything but it's something we should be wary of, especially if we end up generating any additional memory allocators (to do: memory planning with dynamic shapes...)

desertfire commented on test/cpp/jit/test_memory_planning.cpp:
```diff
@@ -0,0 +1,48 @@
+  constexpr int batch_size = 4;
+  constexpr int input_size = 256;
+  int hidden_size = 2 * input_size;
+  auto input = at::randn({batch_size, input_size}, at::kCPU);
```
Can we add a test for CUDA?

makslevental commented on torch/csrc/jit/passes/memory_planning.cpp:
```diff
@@ -0,0 +1,159 @@
+  std::vector<int64_t> sizes, strides;
+  std::tie(sizes, strides) = getSizesStrides(ttp);
+  return at::detail::computeStorageNbytes(
+      sizes, strides, elementSize(ttp->scalarType().value()));
```
one context/general understanding question I had here: what exactly is a strided tensor? in my imagination a strided tensor would be some sparse representation or a contiguous tensor? If that's accurate I don't see understand under what circumstances we'd need to allocate new memory for a strided tensor. 

Maybe this is what you're in fact implying here...

eellison replied:
There's a colab at https://github.com/pytorch/pytorch/wiki/Tensor-and-Operator-Basics#learn-about-view-operations

based off of https://ezyang.github.io/stride-visualizer/index.html

There are basically a limited number of strided patterns that occur:
- default contiguous
- channels-last tensors (important for cuda perf) 
- views of tensors which produce contiguous memory (e.g. permute)
- views of tensor which produce non-dense memory

makslevental commented on torch/csrc/jit/passes/memory_planning.cpp:
```diff
@@ -0,0 +1,159 @@
+  auto sizes = ttp->sizes().concrete_sizes().value();
+  std::vector<int64_t> strides;
+  auto _strides = ttp->strides().concrete_sizes();
+  if (_strides.has_value() && _strides.value()[0] != 0) {
```
LTC->TS sometimes produces `strides=[0,0]`, e.g.,

```
  %16 : Float(1, 500, strides=[0, 0]) = aten::addmm(%p2, %13, %9, %14, %15)
  %17 : Float(1, 500, strides=[0, 0]) = aten::relu(%16)
  ```
  
  @Krovatkin reports that it's in cases of "no device" and "no autograd". Not sure why. I've also seen strangeness with views. In my local branch I have `TODO`s to investigate.

eellison replied:
hmmm, strange

makslevental commented on torch/csrc/jit/passes/memory_planning.cpp:
```diff
@@ -0,0 +1,159 @@
+
+void insertAllocTensorNodes(
+    std::shared_ptr<Graph>& graph,
+    std::vector<Node*> out_nodes,
```
`const` correctness is like flossing - no one practices consistently it until right before their yearly check up :)

makslevental commented on torch/csrc/jit/passes/memory_planning.cpp:
```diff
@@ -0,0 +1,159 @@
+    std::shared_ptr<Graph>& graph,
+    std::vector<Node*> out_nodes,
+    Node* storage,
+    std::map<Value*, c10::TensorTypePtr>& managed_tensors) {
```
There's currently no reason (and I will change it right now in case a reason never materializes) but if we want to implement some kind of locality sensitive allocation I think ordering (rather than determinism) might be important there.

eellison replied:
deterministic output is a good reason to use an ordered map, so if making it unordered introduces indeterminism we should use an ordered map

eellison replied:
locality sensitive allocation would be great obviously but i think we would need to add a bunch more logic than exists, or at the very least add comments about why using an ordered map achieves this

makslevental commented on torch/csrc/jit/runtime/register_prim_ops_fulljit.cpp:
```diff
@@ -127,6 +130,48 @@ RegisterOperators reg(
+
+             uint8_t* start = static_cast<uint8_t*>(buffer.data());
+             void* src = static_cast<void*>(start + offset);
+             at::Tensor sub_tensor = at::from_blob(
```
I briefly discussed this with @hlu1 - namely the fact that out variants should be calling resize on output themselves in order to get the striding right. greping the code base I only found calls like `output.resize_({sizeD, osizeT, osizeH, osizeW});` so I wasn't aware some kind of "one size fits all" like `.resize_(0)` (or `.resize_({0})`) does the right thing. It's actually not clear to me what's happening in that snippet since I don't see a second resize (unless you're implying that `.add` has an implicit resize...

eellison replied:
What we want to have happen is:
- allocate tensor of a given size, at an offset in the slab of memory
- resize_(0) so that the output of the node will set the correct strides
- when that resize happens, still use the same offset in memory for the newly resized tensor 

In the example, `y` and `z` are sharing memory. We resize `y` to 0, then we use it as the out of the add. `y` has the correct channels-last striding that we want, and we also show that it is still using the memory offset that it was initialized to when it is resized, because `y.sum() == z.sum()`. When we pass a tensor of the correct size (not resized to 0), we see that the channels-last memory is not preserved, striding is taken from the input.

eellison replied:
I think this is still relevant - but we can leave for a follow up. Maybe add comment of TODO ?

eellison reviewed (commented):
One other thing we need to worry about in the future is ensuring the dtype matches up for all of the out variants:

```
>>> x = torch.rand([2, 2], dtype=torch.double)
>>> y = torch.rand([2, 2], dtype=torch.float)
>>> torch.add(y, y, out=x).dtype
torch.float64
>>> torch.add(y, y).dtype
torch.float32
```

Commit:
Update on "memory planning"


in principle this depends on https://github.com/pytorch/pytorch/pull/62377 but can be used with a shape profiling pass (i.e. as done in the test).

still remaining i have liveness and alias analysis and a sophisticated memory planning algo (currently just adding up all of the sizes). symbolic shapes are also not handled.

currently the test simply tests for exceptions (of which there are none). going forward I will need to think about what's an appropriate test. run with `PYTORCH_JIT_LOG_LEVEL=>>memory_planning` to see some action.

[ghstack-poisoned]

makslevental committed:
Update on "memory planning"


in principle this depends on https://github.com/pytorch/pytorch/pull/62377 but can be used with a shape profiling pass (i.e. as done in the test).

still remaining i have liveness and alias analysis and a sophisticated memory planning algo (currently just adding up all of the sizes). symbolic shapes are also not handled.

currently the test simply tests for exceptions (of which there are none). going forward I will need to think about what's an appropriate test. run with `PYTORCH_JIT_LOG_LEVEL=>>memory_planning` to see some action.

[ghstack-poisoned]

makslevental committed:
Update on "memory planning"


in principle this depends on https://github.com/pytorch/pytorch/pull/62377 but can be used with a shape profiling pass (i.e. as done in the test).

still remaining i have liveness and alias analysis and a sophisticated memory planning algo (currently just adding up all of the sizes). symbolic shapes are also not handled.

currently the test simply tests for exceptions (of which there are none). going forward I will need to think about what's an appropriate test. run with `PYTORCH_JIT_LOG_LEVEL=>>memory_planning` to see some action.

[ghstack-poisoned]

makslevental committed:
Update on "memory planning"


in principle this depends on https://github.com/pytorch/pytorch/pull/62377 but can be used with a shape profiling pass (i.e. as done in the test).

still remaining i have liveness and alias analysis and a sophisticated memory planning algo (currently just adding up all of the sizes). symbolic shapes are also not handled.

currently the test simply tests for exceptions (of which there are none). going forward I will need to think about what's an appropriate test. run with `PYTORCH_JIT_LOG_LEVEL=>>memory_planning` to see some action.

[ghstack-poisoned]

Commit:
Update on "memory planning"


in principle this depends on https://github.com/pytorch/pytorch/pull/62377 but can be used with a shape profiling pass (i.e. as done in the test).

still remaining i have liveness and alias analysis and a sophisticated memory planning algo (currently just adding up all of the sizes). symbolic shapes are also not handled.

currently the test simply tests for exceptions (of which there are none). going forward I will need to think about what's an appropriate test. run with `PYTORCH_JIT_LOG_LEVEL=>>memory_planning` to see some action.

[ghstack-poisoned]

Commit:
Update on "first sketch memory planning"


in principle this depends on https://github.com/pytorch/pytorch/pull/62377 but can be used with a shape profiling pass (i.e. as done in the test).

still remaining i have liveness and alias analysis and a sophisticated memory planning algo (currently just adding up all of the sizes). symbolic shapes are also not handled.

currently the test simply tests for exceptions (of which there are none). going forward I will need to think about what's an appropriate test. run with `PYTORCH_JIT_LOG_LEVEL=>>memory_planning` to see some action.

[ghstack-poisoned]

Commit:
Update on "first sketch memory planning"


in principle this depends on https://github.com/pytorch/pytorch/pull/62377 but can be used with a shape profiling pass (i.e. as done in the test).

still remaining i have liveness and alias analysis and a sophisticated memory planning algo (currently just adding up all of the sizes). symbolic shapes are also not handled.

currently the test simply tests for exceptions (of which there are none). going forward I will need to think about what's an appropriate test. run with `PYTORCH_JIT_LOG_LEVEL=>>memory_planning` to see some action.

[ghstack-poisoned]

Commit:
Update on "[JIT] memory planning with various strategies"


This PR adds static memory planning for traces with shape info. Currently there are three strategies implemented

* Linear scan (a heuristic based on https://www.usenix.org/legacy/events/vee05/full_papers/p132-wimmer.pdf)
* Greedy by size (a heuristic that allocates largest tensors first)
* Greedy by operator breadth (a heuristic that allocates largest tensors for operators with largest breadths first)

The latter 2 are based on https://arxiv.org/pdf/2001.03288.pdf. 

Most of the action is under `torch/csrc/jit/passes/memory_planning/`. 
Currently this only supports CPU. In principle there's nothing blocking supporting GPU (just requires adding cudaAllocator to GetAllocator) but there are lots of implicit assumptions encoded in the allocation strategies that are probably false for GPU.

Currently the tests don't do anything except exercise the code paths. As we find corner cases we will add tests to check for them.

[ghstack-poisoned]

Commit:
Update on "[JIT] memory planning with various strategies"


This PR adds static memory planning for traces with shape info. Currently there are three strategies implemented

* Linear scan (a heuristic based on https://www.usenix.org/legacy/events/vee05/full_papers/p132-wimmer.pdf)
* Greedy by size (a heuristic that allocates largest tensors first)
* Greedy by operator breadth (a heuristic that allocates largest tensors for operators with largest breadths first)

The latter 2 are based on https://arxiv.org/pdf/2001.03288.pdf. 

Most of the action is under `torch/csrc/jit/passes/memory_planning/`. 
Currently this only supports CPU. In principle there's nothing blocking supporting GPU (just requires adding cudaAllocator to GetAllocator) but there are lots of implicit assumptions encoded in the allocation strategies that are probably false for GPU.

Currently the tests don't do anything except exercise the code paths. As we find corner cases we will add tests to check for them.

[ghstack-poisoned]

Commit:
Update on "[JIT] memory planning with various strategies"


This PR adds static memory planning for traces with shape info. Currently there are three strategies implemented

* Linear scan (a heuristic based on https://www.usenix.org/legacy/events/vee05/full_papers/p132-wimmer.pdf)
* Greedy by size (a heuristic that allocates largest tensors first)
* Greedy by operator breadth (a heuristic that allocates largest tensors for operators with largest breadths first)

The latter 2 are based on https://arxiv.org/pdf/2001.03288.pdf. 

Most of the action is under `torch/csrc/jit/passes/memory_planning/`. 
Currently this only supports CPU. In principle there's nothing blocking supporting GPU (just requires adding cudaAllocator to GetAllocator) but there are lots of implicit assumptions encoded in the allocation strategies that are probably false for GPU.

Currently the tests don't do anything except exercise the code paths. As we find corner cases we will add tests to check for them.

[ghstack-poisoned]

eellison commented on torch/csrc/jit/passes/memory_planning.cpp:
```diff
@@ -0,0 +1,273 @@
+void insertAllocTensorNodes(
+    std::shared_ptr<Graph>& graph,
+    Node* storage,
+    std::unordered_map<LiveRange, Region, live_range_hash> allocations,
```
What does LiveRange represent? I saw it it's size_t start, end, but that is not very descriptive. Could you add some comments to its definition as part of this PR?

eellison commented on torch/csrc/jit/passes/memory_planning.cpp:
```diff
@@ -0,0 +1,273 @@
+    auto region = allocations[lvr];
+    auto allocation = item.second;
+
+    // const_cast fishy?
```
const is sooooo annoying, imo it's rarely worth it

eellison commented on torch/csrc/jit/passes/memory_planning.cpp:
```diff
@@ -0,0 +1,273 @@
+    // const_cast fishy?
+    auto node = const_cast<Node*>(allocation->node());
+
+    // the way that this node magically *becomes* the out varaint is simply
```
nit: varaint/ variant

eellison commented on torch/csrc/jit/passes/memory_planning.cpp:
```diff
@@ -0,0 +1,273 @@
+}
+
+bool hasOutVariant(Node* node) {
+  for (const auto& variant : getAllOperatorsFor(node->kind())) {
```
This is not the best way of checking because you *could* have an operator overloaded with different schemas where one of the schemas has an out variant but another does not. I think this is what your `aten::cat` alludes to.
Maybe an easier way of doing this is to add a Tensor argument to the end of the operator, check if `maybeSchema` exists, then check that the last argument name is `out`. Then remove the extra tensor argument. For the tensor value you can just construct a prim::Uninitialized Node and set the output type to tensor and use that.

eellison commented on torch/csrc/jit/passes/memory_planning.cpp:
```diff
@@ -0,0 +1,273 @@
+      auto size = computeStorageSize(*out_v);
+      if (size.has_value() && size.value() > 0) {
+        managed_tensor_values.insert({out_v, size.value()});
+      } else if (isOptimizableContainerType(node)) {
```
What is the logic here ? of the `isOptimizableContainerType` ?

eellison commented on torch/csrc/jit/passes/memory_planning.cpp:
```diff
@@ -0,0 +1,273 @@
+      } else if (isOptimizableContainerType(node)) {
+        leaked_values.insert(out_v);
+      } else {
+        TORCH_WARN(
```
I think is better under a GRAPH_DEBUG than TORCH_WARN

eellison commented on torch/csrc/jit/passes/memory_planning.cpp:
```diff
@@ -0,0 +1,273 @@
+    std::vector<const Node*>,
+    std::unordered_map<const Value*, uint64_t>,
+    std::unordered_map<const Value*, LiveRange>>
+getManagedStuff(std::shared_ptr<Graph>& graph) {
```
could we rename getManagedStuff lol

eellison replied:
same here about sentence or two description

eellison commented on torch/csrc/jit/passes/memory_planning.cpp:
```diff
@@ -0,0 +1,273 @@
+  return std::make_pair(out_nodes, managed_tensor_values);
+}
+
+std::tuple<
```
Can we use a struct here ? you pretty much never want tuple > 2

eellison commented on torch/csrc/jit/passes/memory_planning.cpp:
```diff
@@ -0,0 +1,273 @@
+}
+
+std::pair<std::vector<const Node*>, std::unordered_map<const Value*, uint64_t>>
+getManagedValues(
```
could you add a sentence or two description above the function what this is achieving?

eellison commented on torch/csrc/jit/passes/memory_planning.cpp:
```diff
@@ -0,0 +1,273 @@
+  std::map<LiveRange, const Value*, live_range_start_cmp> managed_range_values;
+  for (const auto& item : managed_value_ranges) {
+    if (managed_range_values.count(item.second)) {
+      TORCH_WARN(
```
what is this checking ? It seems like this should be an error maybe ?

eellison reviewed (commented):
Nice, looks great !! A few comments for my own understanding / readability

eellison commented on torch/csrc/jit/runtime/register_prim_ops_fulljit.cpp:
```diff
@@ -127,6 +128,48 @@ RegisterOperators reg(
+                 total_size,
+                 std::move(buffer),
+                 /*allocator=*/nullptr,
+                 /*resizable=*/false);
```
what is the resizable false mean here ? could you add a comment ? i guess this is because we want to exactly model the size
